### PR TITLE
build, refactor: bump to C++17

### LINKED ISSUE #1378: Switch to C++17
This is more of a maybe-next-year kind of thing, but I propose we switch from C++11 to C++14 or even C++17 to be able to utilize new functionality and optimizations. Some things we would directly benefit from are:

string_view (C++17)
----------------------------
Allows you to view a part of a string instead of making a copy. Replaces many usages of substr which speeds up boincblock unpacking and `ExtractXML`.

- gcc-7
- clang-4

std::optional (C++17)
----------------------------
Replaces the need to manually carry an `is-initialized` flag, see `StructCPID`.

- gcc-7
- clang-4

std::filesystem (C++17)
----------------------------
Replaces `boost::filesystem`.

- gcc-8
- clang-7

Splicing for maps and sets
-----------------------------------
Cheaper data move between containers.

- gcc-7
- clang-8

std::variant (C++17)
----------------------------
Replaces `boost::variant`.

- gcc-7
- clang-4

Relaxing constraints on constexpr functions (C++14)
----------------------------------------------------------
Allows more flexibility in constexpr functions which gives the compiler a better chance at optimizing.

- ??

STL constexpr improvements (C++14)
---------------------------------------------------
Several functions in the STL have been made constexpr. 

- gcc-5
- clang-3.4

Parallel algorithms (C++14)
----------------------------
Allows for standard algorithm (find, sort etc) to utilize more cores.

- Not yet supported in gcc or clang

### PR DESCRIPTION
**Pull request by div72**

### CODE REVIEW

**Comment by barton2526** [comment]
Linux build fails because Actions is using Ubuntu 16
```FILE_ENV: ./ci/test/00_setup_env_native_old.sh```

`Creating ubuntu:16.04 container to run in`

**Comment by barton2526** [comment]
@div72 `std::random_shuffle` is deprecated and replaced by `std::shuffle`

**Comment by denravonska** [comment]
`boost::optional` and `boost::any` can now be replaced with `std::` versions. Also, `boost::variant` if it's used.

File: src/compat/assumptions.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+#ifndef BITCOIN_COMPAT_ASSUMPTIONS_H
+#define BITCOIN_COMPAT_ASSUMPTIONS_H
+
+#include <limits>
+
+// Assumption: We assume that the macro NDEBUG is not defined.
+// Example(s): We use assert(...) extensively with the assumption of it never
+//             being a noop at runtime.
+#if defined(NDEBUG)
+# error "Bitcoin cannot be compiled without assertions."
```
Comment (div72) [inline src/compat/assumptions.h:17]: I am conflicted on changing this message. My reasoning behind not changing it is to make it easier to cherry-pick commits involving this file as this text will only be present when compiling and won't be seen in the wallet itself. Still open for discussion.

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -1532,7 +1531,16 @@ bool CWallet::SelectCoinsMinConf(int64_t nTargetValue, unsigned int nSpendTime,
...
+    // A wrapper for GetRandInt to satisfy std::shuffle's UniformRandomBitGenerator requirements.
+    // TODO: Port FastRandomContext and replace this.
+    class {
+    public:
+        typedef int result_type;
+        static constexpr int min() { return 0; }
+        static constexpr uint64_t max() { return 0x7FFFFFFF; }
+        inline uint64_t operator()() noexcept { return GetRandInt(0x7FFFFFFF); }
+    } urbg;
+    std::shuffle(vCoins.begin(), vCoins.end(), urbg);
```
Comment (div72) [inline src/wallet/wallet.cpp]: This is a really ugly workaround, will address this in another PR.

**Review by div72 — COMMENTED** [review]
This PR lacks changes for `boost::chrono`, `boost::filesystem` and `boost::thread`. I will address `boost::chrono` and `boost::thread` in another PR and wait for upstream for `boost::filesystem`.

File: src/compat/assumptions.h (hunk 1 above)
Comment (jamescowens) [reply src/compat/assumptions.h:17]: Leave it is my vote.

**Comment by div72** [comment]
@cyrossignol When you have some free time, could you take a look at this on macOS? The CI seems to hang forever but I am unable to reproduce it with clang on linux.

File: src/compat/assumptions.h (hunk 1 above)
Comment (cyrossignol) [reply src/compat/assumptions.h:17]: Agreed.

**Comment by cyrossignol** [comment]
⏳

File: src/wallet/wallet.cpp (hunk 2 above)
Comment (jamescowens) [reply src/wallet/wallet.cpp]: Ugh. Yep.

Comment (cyrossignol) [reply src/wallet/wallet.cpp]: @div72 The macOS tests are stalling on `std::shuffle()` here when `wallet_tests::coin_selection_tests()` calls `CWallet::SelectCoinsMinConf()`. Not sure why yet, but gdb seems to indicate that something in the linked Homebrew OpenSSL's libcrypto 1.1.0 is hitting an infinite loop when initializing some state for the `GetRandInt()` call. 

I'll try to come back to this tonight when I'll have more time. Really hope it's not an OpenSSL issue...

Comment (cyrossignol) [reply src/wallet/wallet.cpp]: Can't unresolve. Don't think I have any repo permissions.

Comment (div72) [reply src/wallet/wallet.cpp]: @cyrossignol Of course this had to be faulty, isn't it?

Are you sure the problem is OpenSSL's? Could it be related to https://stackoverflow.com/questions/22915325/avoiding-self-assignment-in-stdshuffle?

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -1520,6 +1519,19 @@ struct smallestcoincomp
...
+    while (first != last) {
+        uint64_t j = GetRand(last - first);
+        if (j) {
+            using std::swap;
+            swap(*first, *(first + j));
+        }
+        ++first;
+    }
+}
+
```
Comment (jamescowens) [inline src/wallet/wallet.cpp]: Sigh... sometimes I feel like progress is not progress with the new C++ standards...

Comment (div72) [reply src/wallet/wallet.cpp]: Also one of the reasons why I don't like C++.

Then again just std::shuffle with the stdlib's random could be used and that might work but I don't exactly know if that'll create any issues.

Comment (jamescowens) [reply src/wallet/wallet.cpp]: I think the std::shuffle with std::default_random_engine(seed) as the generator will work, right? The seed can use the epoch time.

Comment (jamescowens) [reply src/wallet/wallet.cpp]: There is nothing particularly sensitive about this shuffle from a wallet point of view as long as it generates a different shuffle order each time it is called.

Comment (div72) [reply src/wallet/wallet.cpp]: Using `GetAdjustedTime()` as seed fails wallet_tests as the tests are run so fast that the time doesn't change. Using `static_cast<unsigned int>(GetRand(0x80000000))` as seed works but it is a little bit weird imo.

Comment (jamescowens) [reply src/wallet/wallet.cpp]: GetTimeMillis(), the same medium resolution timer I use for the MilliTimer?

Comment (jamescowens) [reply src/wallet/wallet.cpp]: Or std::chrono::milliseconds GetTime() if you don't like the underlying use of boost's time classes.

Comment (jamescowens) [reply src/wallet/wallet.cpp]: There is a microsecond level too if we need it.

**Review by jamescowens — APPROVED** [review]
utACK

**Comment by iFoggz** [comment]
utACK
nice to see c++17 coming in and boost going away.

**Review by cyrossignol — APPROVED** [review]
Looks like macOS issues resolved.

**Pull request merged**